src/pages/images/actions/UploadInstanceBtn.tsx

```diff
@@ -0,0 +1,230 @@
+import { ChangeEvent, FC, useState } from "react";
+import {
+  ActionButton,
+  Button,
+  Input,
+  Modal,
+} from "@canonical/react-components";
+import { Form, useNavigate } from "react-router-dom";
+import { useFormik } from "formik";
+import * as Yup from "yup";
+import { instanceNameValidation } from "util/instances";
+import { useProject } from "context/project";
+import { useQueryClient } from "@tanstack/react-query";
+import { queryKeys } from "util/queryKeys";
+import { uploadInstance } from "api/instances";
+import { UploadState } from "types/storage";
+import { useEventQueue } from "context/eventQueue";
+import { useToastNotification } from "context/toastNotificationProvider";
+import ProgressBar from "components/ProgressBar";
+import { humanFileSize } from "util/helpers";
+import usePortal from "react-useportal";
+import StoragePoolSelector from "pages/storage/StoragePoolSelector";
+
+export interface UploadInstanceFormValues {
+  instanceFile: File | null;
+  name: string;
+  pool: string;
+}
+
+const UploadInstanceBtn: FC = () => {
+  const { openPortal, closePortal, isOpen, Portal } = usePortal();
+  const { project, isLoading: isProjectLoading } = useProject();
+  const [isUploading, setIsUploading] = useState(false);
+  const [uploadState, setUploadState] = useState<UploadState | null>(null);
+  const eventQueue = useEventQueue();
+  const toastNotify = useToastNotification();
+  const queryClient = useQueryClient();
+  const navigate = useNavigate();
+  const [uploadAbort, setUploadAbort] = useState<AbortController | null>(null);
+  const instanceNameAbort = useState<AbortController | null>(null);
+
+  const changeFile = (e: ChangeEvent<HTMLInputElement>) => {
+    const { onChange } = formik.getFieldProps("instanceFile");
+    onChange(e);
+
+    if (e.currentTarget.files) {
+      const file = e.currentTarget.files[0];
+      // remove file extension
+      const instanceName = file.name.split(".tar")[0];
+      const instanceNameSuffix = "-new";
+      void formik.setFieldValue("name", instanceName + instanceNameSuffix);
+      void formik.setFieldValue("instanceFile", file);
+    }
+  };
+
+  const handleCancelUpload = () => {
+    uploadAbort?.abort();
+    closePortal();
+  };
+
+  const handleUpload = async () => {
+    if (!formik.values.instanceFile) {
+      return;
+    }
+
+    setIsUploading(true);
+    const uploadController = new AbortController();
+    setUploadAbort(uploadController);
+
+    try {
+      const instanceName = formik.values.name;
+      const operation = await uploadInstance(
+        formik.values.instanceFile,
+        instanceName,
+        project?.name,
+        formik.values.pool,
+        setUploadState,
+        uploadController,
+      );
```

Comment:
Sure, it's a little bit of a personal preference to avoid promise chains, but you are right about consistency and I don't think we should break that in this PR. Will adjust :+1: